addons/point_of_sale/static/src/js/models.js

```diff
@@ -746,7 +748,11 @@ exports.PosModel = Backbone.Model.extend({
         var self = this;
         this.set('synch',{ state: 'connecting', pending: orders.length});
 
-        return self._save_to_server(orders, options).done(function (server_ids) {
+        // Need to shallow copy orders to avoid deleting new orders (which are saved to "orders" in localStorage also)
+        // while previous orders are in progress of "_save_to_order" and removed them after "_save_to_order" successfull
+        var flush_orders = Object.assign([], orders);
+
+        return self._save_to_server(flush_orders, options).done(function (server_ids) {
```

Comment:
Ideally I think the server should return references to the orders that have actually been saved. It kind of does this, but it returns the pos.order record IDs which isn't particularly useful because we don't know what database ID corresponds with what JSON order. It would have been nicer if it returned instead the ```name``` property of the synced JSON orders. Then we could simply use that in the callback of ```posOrderModel.call('create_from_ui')``` to only delete orders actually synced. We cannot make such a change in a stable release though. In theory we could do another RPC call in JS to read the ```pos_reference``` (= JSON ```name```) fields of those pos.orders. But I like to avoid additional RPC if possible.

So instead how about the following:

Before the RPC call to ```create_from_ui``` we save the ```id``` properties of the objects in the ```orders``` array:

```javascript
var order_ids = _.pluck(orders, 'id');
```

After this we use that array in the callback when deleting orders

```javascript
_.each(order_ids, function (id) {
    self.db.remove_order(id);
});
```

I think this should resolve the problem since I think that the issue occurs in the time between the ```create_from_ui``` call and the callback right? Don't hesitate to point out if I'm missing something :).